### PR TITLE
Refactor fracture domain integrals to use vectorpostprocessors

### LINKED ISSUE #9559: Fracture Domain Integral Cleanup
### Description of the enhancement or error report
We need to do the following to clean up the fracture domain integrals:
- [x] Remove the requirement that the q function be stored as an AuxVariable for every node on the crack front.  This should still be an option to help debug the model, but not the default behavior.
- [ ] Convert the fracture integral Postprocessors to VectorPostprocessors
- [x] Get rid of the material models that compute auxiliary fields and compute those quantities on the fly in the interaction integrals.
- [x] Enable the usage of fracture integrals with TensorMechanics

### Rationale for the enhancement or information for reproducing the error
In large 3D models with many points along the crack front, there can be a very large number of AuxVariables to store the q function.  Also, the fracture integrals are currently not flexible to allow for evolving cracks because of the reliance on AuxVariables for the q functions and their gradients.  

The fracture integrals were originally written before VectorPostprocessors existed, and the code and output can be cleaned up quite a bit by eliminating the individual Postprocessors for each point on the crack front.  That will also make the code more flexible to allow for evolving crack fronts.

The usage of material properties for the interaction integrals significantly slows down the computation because those properties get evaluated at every iteration.

We need to move these to TensorMechanics like everything else.

### Identified impact
Changes to classes used by domain integrals: JIntegral, InteractionIntegral postprocessors, CrackFrontDefinition, DomainIntegral action, AuxKernels for computing q functions.

### PR DESCRIPTION
**Pull request by bwspenc**
closes #9559 

The system for fracture integrals was developed before VectorPostprocessors (VPPs) were introduced in MOOSE. These integrals involve computations of quantities along lines on the the crack front, so the VPP system is a much more natural way to do this than the current way, which creates a (large!) set of Postprocessors.

### CODE REVIEW

File: modules/solid_mechanics/include/vectorpostprocessors/InteractionIntegralSM.h [hunk 1]
```diff
@@ -76,6 +76,14 @@ class InteractionIntegralSM : public ElementIntegralPostprocessor
 
   const QMethod _q_function_type;
 
+  enum class PositionType
+  {
+    Angle,
+    Distance
+  };
+
+  const PositionType _position_type;
```
Comment (dschwen) [inline modules/solid_mechanics/include/vectorpostprocessors/InteractionIntegralSM.h]: those two statements could be combined into one

File: modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C [hunk 2]
```diff
@@ -123,42 +132,33 @@ InteractionIntegralSM::initialSetup()
...
-      grad_q(j) += dphi_curr_elem[i][_qp](j) * _q_curr_elem[i];
-  }
-
   // In the crack front coordinate system, the crack direction is (1,0,0)
-  RealVectorValue crack_direction(0.0);
-  crack_direction(0) = 1.0;
+  RealVectorValue crack_direction(1.0, 0.0, 0.0);
 
   // Calculate (r,theta) position of qp relative to crack front
   Point p(_q_point[_qp]);
```
Comment (dschwen) [inline modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: Why is a copy of `_q_point[_qp]` created here?

File: modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C [hunk 3]
```diff
@@ -263,27 +259,81 @@ InteractionIntegralSM::computeIntegral()
   fe->reinit(_current_elem);
 
   // calculate q for all nodes in this element
-  _q_curr_elem.clear();
   unsigned int ring_base = (_q_function_type == QMethod::Topology) ? 0 : 1;
 
-  for (unsigned int i = 0; i < _current_elem->n_nodes(); ++i)
+  for (unsigned int icfp = 0; icfp < _interaction_integral.size(); icfp++)
```
Comment (dschwen) [inline modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: I thought we were moving away form `unsigned int` loop indices. At the very least I'd use `std::size_t`

File: modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C [hunk 4]
```diff
@@ -263,27 +259,81 @@ InteractionIntegralSM::computeIntegral()
...
+
+void
+InteractionIntegralSM::threadJoin(const UserObject & y)
+{
+  const InteractionIntegralSM & uo = static_cast<const InteractionIntegralSM &>(y);
 
-  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-    sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
-  return sum;
+  for (auto i = beginIndex(_interaction_integral); i < _interaction_integral.size(); ++i)
```
Comment (dschwen) [inline modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: or, like here, beginIndex

File: modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C [hunk 5]
```diff
@@ -106,8 +114,9 @@ InteractionIntegralSM::InteractionIntegralSM(const InputParameters & parameters)
 
   // Checking for consistency between mesh size and length of the provided displacements vector
   if (_ndisp != _mesh.dimension())
-    mooseError(
-        "The number of variables supplied in 'displacements' must match the mesh dimension.");
+    mooseError("InteractionIntegralSM Error: number of variables supplied in 'displacements' must "
+               "match the mesh dimension.");
```
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: The problem with paramError is that it puts a little string on the end (can't remember what it is) that doesn't really make sense after this sentence.

[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: I just tested this.  I'm not crazy.  It tacks this 'not present in InputParams' to the end of the string in line 877 of InputParameters.h:
```
*** ERROR ***
param 'DomainIntegral error: must set radius_inner and radius_outer.' not present in InputParams

[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: Looks like if you use the version of `paramError()` that takes two arguments, you get a better looking message. That version only makes sense when it's related to only one parameter, though.

**Comment by bwspenc** [comment]
@dschwen I just pushed an update that I believe addresses all of the issues you pointed out in the files that I touched. All except converting `mooseError()` to `paramError()`, that is. That one requires some more investigation.

File: modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C (hunk 5 above)
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: It doesn't tack a message on! And there is no single argument version of `paramError`. The function is always interpreting the first argument as the parameter name - which it takes to be `DomainIntegral error: must set radius_inner and radius_outer.` in your example - which would be a ridiculous name for a parameter, so yeah... it obviously is not present in InputParams :-)

The case I pointed out is really only related to a single parameter: `displacements` (which I actually gave an example for in my comment above...

Your point about combinations of multiple parameters throwing an error is taken though. I've encountered that before and always tried to pick one of the parameters to tag with the error.

[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/solid_mechanics/src/vectorpostprocessors/InteractionIntegralSM.C]: I just pushed an update that converts the instances of mooseError that make sense to paramError.

**Review by dschwen — APPROVED** [review]

**Pull request merged**